File: backend/src/modules/auth/auth.service.ts
import { User, Role, UserRole, Permission } from '@database/index';
import { hashPassword, comparePassword } from '@utils/password';
import { signAccessToken, signRefreshToken, verifyRefreshToken } from '@utils/jwt';
import { AppError } from '@utils/AppError';
import { USER_STATUS } from '@config/constants';
import { AuthResult, TokenPair } from './auth.types';
import {
  RegisterInput,
  LoginInput,
  RefreshInput,
} from './auth.validation';

function buildTokenPair(userId: string): TokenPair {
  return {
    accessToken: signAccessToken(userId),
    refreshToken: signRefreshToken(userId),
  };
}

export async function register(input: RegisterInput): Promise<AuthResult> {
  const existing = await User.findOne({ where: { email: input.email } });
  if (existing) {
    throw AppError.conflict('Email is already registered');
  }

  const passwordHash = await hashPassword(input.password);
  const user = await User.create({
    name: input.name,
    email: input.email,
    passwordHash,
    status: USER_STATUS.ACTIVE,
  });

  const userRole = await Role.findOne({ where: { slug: 'user' } });
  if (userRole) {
    await UserRole.create({ userId: user.id, roleId: userRole.id });
  }

  const tokens = buildTokenPair(String(user.id));

  return {
    ...tokens,
    user: { id: user.id, name: user.name, email: user.email },
  };
}

export async function login(input: LoginInput): Promise<AuthResult> {
  const user = await User.scope('withPassword').findOne({ where: { email: input.email } });
  if (!user) {
    throw AppError.unauthorized('Invalid email or password');
  }

  if (user.status !== USER_STATUS.ACTIVE) {
    throw AppError.forbidden('Account is not active');
  }

  const valid = await comparePassword(input.password, user.passwordHash);
  if (!valid) {
    throw AppError.unauthorized('Invalid email or password');
  }

  user.lastLoginAt = new Date();
  await user.save({ fields: ['lastLoginAt'] });

  const tokens = buildTokenPair(String(user.id));

  return {
    ...tokens,
    user: { id: user.id, name: user.name, email: user.email },
  };
}

export async function refresh(input: RefreshInput): Promise<TokenPair> {
  let payload;
  try {
    payload = verifyRefreshToken(input.refreshToken);
  } catch {
    throw AppError.unauthorized('Invalid or expired refresh token');
  }

  if (payload.type !== 'refresh') {
    throw AppError.unauthorized('Invalid token type');
  }

  const user = await User.findByPk(payload.sub);
  if (!user || user.status !== USER_STATUS.ACTIVE) {
    throw AppError.unauthorized('Invalid or expired refresh token');
  }

  return buildTokenPair(String(user.id));
}

export async function getMe(userId: string) {
  const user = await User.findByPk(userId, {
    attributes: ['id', 'name', 'email', 'avatar', 'bio', 'status', 'createdAt'],
    include: [
      {
        model: Role,
        as: 'roles',
        attributes: ['id', 'name', 'slug'],
        through: { attributes: [] },
        include: [
          {
            model: Permission,
            as: 'permissions',
            attributes: ['id', 'name', 'slug', 'module', 'description'],
            through: { attributes: [] },
          },
        ],
      },
    ],
  });
  if (!user) {
    throw AppError.notFound('User not found');
  }
  return user;
}

File: backend/app/services/auth_service.py
from datetime import datetime

from app.config import Const
from app.core import security
from app.core.errors import AppError
from app.models import Role, User, UserRole


def _token_pair(user_id):
    return {
        'accessToken': security.sign_access_token(user_id),
        'refreshToken': security.sign_refresh_token(user_id),
    }


def _auth_user(user):
    return {'id': user.id, 'name': user.name, 'email': user.email}


def register(db, data):
    existing = db.query(User).filter(User.email == data['email']).first()
    if existing:
        raise AppError.conflict('Email is already registered')

    user = User(
        name=data['name'],
        email=data['email'],
        passwordHash=security.hash_password(data['password']),
        status=Const.USER_ACTIVE,
    )
    db.add(user)
    db.flush()

    user_role = db.query(Role).filter(Role.slug == 'user').first()
    if user_role:
        db.add(UserRole(userId=user.id, roleId=user_role.id))

    db.commit()
    db.refresh(user)

    return {**_token_pair(user.id), 'user': _auth_user(user)}


def login(db, data):
    user = db.query(User).filter(User.email == data['email']).first()
    if not user:
        raise AppError.unauthorized('Invalid email or password')
    if user.status != Const.USER_ACTIVE:
        raise AppError.forbidden('Account is not active')
    if not security.verify_password(data['password'], user.passwordHash):
        raise AppError.unauthorized('Invalid email or password')

    user.lastLoginAt = datetime.utcnow()
    db.commit()
    db.refresh(user)

    return {**_token_pair(user.id), 'user': _auth_user(user)}


def refresh(db, data):
    try:
        payload = security.verify_refresh_token(data['refreshToken'])
    except Exception:
        raise AppError.unauthorized('Invalid or expired refresh token')

    if payload.get('type') != 'refresh':
        raise AppError.unauthorized('Invalid token type')

    user = db.query(User).filter(User.id == int(payload['sub']), User.deletedAt.is_(None)).first()
    if not user or user.status != Const.USER_ACTIVE:
        raise AppError.unauthorized('Invalid or expired refresh token')

    return _token_pair(user.id)


def get_me(db, user_id):
    user = db.query(User).filter(User.id == user_id, User.deletedAt.is_(None)).first()
    if not user:
        raise AppError.not_found('User not found')
    return user